triton-cli/src/main.rs

```diff
@@ -0,0 +1,259 @@
+    format!("0x{hex}")
+}
+
+fn verify(proof_path: &String) {
+    let (stark, claim, proof) = read_proof(proof_path).expect("Failed to load proof");
+
+    let verdict = triton_vm::verify(stark, &claim, &proof);
+    if !verdict {
+        println!("Proof is not valid!");
+        return;
+    }
+    println!("proof is valid!");
+    println!("program digest: {}", digest_to_str(claim.program_digest));
+    println!("=================");
+    println!("security level: {} bits", stark.security_level);
+    println!("FRI expansion factor: {}", stark.fri_expansion_factor);
+    println!("trace randomizers: {}", stark.num_trace_randomizers);
+    println!("colinearity checks: {}", stark.num_collinearity_checks);
+    println!("codeword checks: {}", stark.num_combination_codeword_checks);
+    println!("=================");
+    println!("public inputs:");
+    if claim.input.is_empty() {
+        println!("(none)");
+    }
+    for v in claim.input {
+        println!("{}", u64::from(v));
+    }
+    println!("public outputs:");
+    if claim.output.is_empty() {
+        println!("(none)");
+    }
+    for v in claim.output {
+        println!("{}", u64::from(v));
+    }
+}
+
+fn parse_inputs(inputs: Option<String>) -> Vec<BFieldElement> {
+    inputs
+        .unwrap_or_default()
+        .split(',')
+        .filter(|v| !v.is_empty())
+        .map(|v| BFieldElement::from(v.parse::<u64>().unwrap()))
+        .collect::<Vec<BFieldElement>>()
+}
+
+fn prove(
+    asm: &String,
+    out: &String,
+    public_inputs: Option<String>,
+    private_inputs: Option<String>,
+) {
+    // TODO: test paths before making the proof?
+    let asm = fs::read_to_string(asm).expect("unable to read file");
+    let instructions = triton_vm::parser::parse(&asm).unwrap();
+    let l_instructions = triton_vm::parser::to_labelled_instructions(instructions.as_slice());
+    let program = triton_vm::program::Program::new(l_instructions.as_slice());
+
+    let public_input = PublicInput::from(parse_inputs(public_inputs));
+    let non_determinism = NonDeterminism::from(parse_inputs(private_inputs));
+    println!("proving...");
+    let data = triton_vm::prove_program(&program, public_input, non_determinism).unwrap();
+    println!("success!");
+
+    // write the proof in an arbitrary binary format
+    //
+    // version: u8
+    //
+    // security_level: u64
+    // fri_expansion_factor: u64
+    // num_trace_randomizers: u64
+    // num_colinearity_checks: u64
+    // num_combination_codeword_checks: u64
+    //
+    // digest_length: u8
+    // public_input_length: u64
+    // public_output_length: u64
+    // proof_length: u64
+    //
+    // program_digest: u64[digest_length]
+    //
+    // public_inputs: u64[public_input_length]
+    // public_outputs: u64[public_output_length]
+    // proof: u64[proof_length]
+    write_proof(data, out).expect("failed to write proof to file");
+    println!("proof written to: {out}");
+}
```

Comment:
Instead of (de)serializing ad-hoc, how about using libraries built for exactly that purpose instead? In the example below, I'm using `serde` and `bincode`, but other alternatives might exist.

```rust
#[derive(Debug, Clone, Eq, PartialEq, Serialize, Deserialize)]
struct SerializableProof {
    version: u8,
    stark: Stark,
    claim: Claim,
    proof: Proof,
}

fn write_proof(proof_path: &str, (stark, claim, proof): (Stark, Claim, Proof)) -> Result<()> {
    let serializable_proof = SerializableProof {
        version: 1,
        stark,
        claim,
        proof,
    };
    let encoded = bincode::serialize(&serializable_proof)?;
    fs::write(proof_path, &encoded)?;
    Ok(())
}

fn read_proof(proof_path: &str) -> Result<(Stark, Claim, Proof)> {
    let encoded = fs::read(proof_path)?;
    let SerializableProof {
        version,
        stark,
        claim,
        proof,
    } = bincode::deserialize(&encoded)?;
    if version != 1 {
        bail!("unsupported proof version: {version}"); // uses crate `anyhow`
    }
    Ok((stark, claim, proof))
}
```

This is probably not the final version of the code yet: in the suggestion, the proof's version is embedded in the struct `SerializableProof`. However, in order to actually support different proof versions, it should probably be parsed first, and subsequent parsing should then depend on the version.

Reply:
> This is probably not the final version of the code yet: in the suggestion, the proof's version is embedded in the struct SerializableProof. However, in order to actually support different proof versions, it should probably be parsed first, and subsequent parsing should then depend on the version.

Yes, i really just wanted to reserve the first byte so in the future it's possible to implement versioning.

I didn't use a library because i'm not familiar with them. In the example above, will changes to the `Stark`, `Claim`, or `Proof` struct (or any nested structs) change the file structure? I'm not opposed to using libraries but I think it's important that the file structure is well documented/stable so others can write parsers in other languages. Explicitly writing the file byte by byte is an easy way to do that.

Reply:
I guess my question is, how do we make sure `version` is changed anytime the `Stark`/`Claim`/`Proof` struct is modified?

Reply:
I ended up making a custom struct and using serde+bincode with that.

Reply:
> I guess my question is, how do we make sure `version` is changed anytime the `Stark`/`Claim`/`Proof` struct is modified?

Yeah, that's a super valid question. My proposed “solution” falls short on at least two counts:
- since the `version` is a field on the struct itself, making de-serialization conditional to the version doesn't work
- it's not obvious (to me) how to ensure correct versioning if any of `Stark`, `Claim`, or `Proof`'s format changes

A custom (de)serializer certainly allows addressing both of those shortcomings. That said, it feels as if this is a problem other people have encountered before, and have provided a general solution for. Maybe this solution is `serde` ∧/∨ `bincode`, but I'm not sure. I'll take another look. :relieved:

Reply:
> Yes, i really just wanted to reserve the first byte so in the future it's possible to implement versioning.

That is a super good idea as far as I can see. :+1: